### PR TITLE
Enable leak_detection for "fault_injection" test case.

### PR DESCRIPTION
**Pull request by saxena-anurag**
## Description

This PR:
1. enables "leak_detection" for `fault_injection` test case for `unit_tests.exe`
2. Fixes test bugs to allow leak_detection to work.
3. Fixes leaks found in product code.

## Testing

This PR adds leak_detection test.

## Documentation

NA

### CODE REVIEW

File: tests/end_to_end/helpers.h [hunk 1]
```diff
@@ -32,6 +32,22 @@ typedef struct _ebpf_free_memory
 
 typedef std::unique_ptr<uint8_t, ebpf_free_memory_t> ebpf_memory_t;
 
+typedef struct _close_bpf_link
+{
+    void
+    operator()(bpf_link* link)
```
Comment (dthaler) [inline tests/end_to_end/helpers.h:38]:
```suggestion
    operator()(_In_opt_ bpf_link* link)
```
add SAL?

Comment (saxena-anurag) [reply tests/end_to_end/helpers.h]: Fixed.

File: libs/platform/unit/platform_unit_test.cpp [hunk 2]
```diff
@@ -485,25 +504,31 @@ TEST_CASE("extension_test", "[platform]")
...
+        ebpf_provider_unload(provider_context);
+    }
+    REQUIRE(result == EBPF_SUCCESS);
 
     REQUIRE(returned_provider_data == &provider_data);
     REQUIRE(returned_provider_dispatch_table == &test_provider_dispatch_table);
 
     ebpf_extension_unload(client_context);
+#pragma warning(push)
+#pragma warning(disable : 6001) // Using uninitialized memory 'provider_context'.
```
Comment (dthaler) [inline libs/platform/unit/platform_unit_test.cpp:529]: Since line 483 actually initializes it, why is this needed?  Do we think there is a bug in the compiler's checker?

**Review by dthaler — APPROVED** [review]

File: libs/platform/unit/platform_unit_test.cpp (hunk 2 above)
Comment (saxena-anurag) [reply libs/platform/unit/platform_unit_test.cpp:529]: Yes I was not able to figure out why compiler was flagging this. One possible reason may be due to REQUIRE() statements, and static analyzer is somehow not able to determine if `provider_context` was initialized.

File: tests/end_to_end/end_to_end.cpp [hunk 3]
```diff
@@ -376,26 +390,26 @@ droppacket_test(ebpf_execution_type_t execution_type)
...
+    bpf_link_ptr link;
 
     single_instance_hook_t hook(EBPF_PROGRAM_TYPE_XDP, EBPF_ATTACH_TYPE_XDP);
     program_info_provider_t xdp_program_info(EBPF_PROGRAM_TYPE_XDP);
 
     const char* file_name = (execution_type == EBPF_EXECUTION_NATIVE ? "droppacket_um.dll" : "droppacket.o");
-
-    result = ebpf_program_load(file_name, BPF_PROG_TYPE_UNSPEC, execution_type, &object, &program_fd, &error_message);
+    result =
+        ebpf_program_load(file_name, BPF_PROG_TYPE_UNSPEC, execution_type, &unique_object, &program_fd, &error_message);
```
Comment (dv-msft) [inline tests/end_to_end/end_to_end.cpp:402]: If we revisit this call's implementation in future, could we should consider converting the return value type to that of an ```std::variant``` of the error code and 'unique_object'?  We could also convert the 'error_message' argument to an ```std::optional```.

(The current implementation is fine for now)

File: tests/end_to_end/end_to_end.cpp [hunk 4]
```diff
@@ -522,10 +533,9 @@ divide_by_zero_test_um(ebpf_execution_type_t execution_type)
 
     REQUIRE(hook_result == 0);
 
-    hook.detach_link(link);
-    hook.close_link(link);
+    hook.detach_and_close_link(&link);
 
-    bpf_object__close(object);
+    bpf_object__close(unique_object.release());
```
Comment (dv-msft) [inline tests/end_to_end/end_to_end.cpp:538]: Genuine question: why couldn't we use a custom deleter for this unique ptr to do the close for us at its destruction?
(The current implementation is fine. Even if a custom deleter _is_ possible, we can look at it in the next implementation update)

Comment (saxena-anurag) [reply tests/end_to_end/end_to_end.cpp:538]: We are using custom deleter for this unique_ptr (see line 78 in this file). In the case the test bails due to a failure, custom deleter will clean up the `bpf_object`.

I agree we can completely remove line 538 and it will still work, but I wanted to keep the flow similar to what it was earlier.

**Review by dv-msft — APPROVED** [review]
